tests/gui/test_helper_find.py

```diff
@@ -0,0 +1,145 @@
+# ###################################################
+# Copyright (C) 2008-2016 The Unknown Horizons Team
+# team@unknown-horizons.org
+# This file is part of Unknown Horizons.
+#
+# Unknown Horizons is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the
+# Free Software Foundation, Inc.,
+# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+# ###################################################
+
+import unittest
+
+import mock
+
+from tests.gui.helper import GuiHelper
+
+
+class Widget(object):
+	"""
+	Fake widget with a name, children and possible a parent. All children of this widget will
+	get their parent attribute set to this widget.
+	"""
+	def __init__(self, name, children=None):
+		self.name = name
+		self.parent = None
+		self.children = children or []
+		for c in self.children:
+			c.parent = self
+
+	def __hash__(self):
+		return id(self)
+
+
+class FindWidgetTest(unittest.TestCase):
+	"""
+	Tests about finding widgets in pychan's internal lists.
+	"""
+	def setUp(self):
+		self.widgets = set()
+
+		def w(name, children=None):
+			widget = Widget(name, children)
+			self.widgets.add(widget)
+			return widget
+
+		self.w = w
+
+		helper = GuiHelper(mock.Mock(), mock.Mock())
+		self.find = lambda x: helper._find(self.widgets, x)
+
+	def test_simple(self):
+		"""
+		Single widget in the global list.
+		"""
+		target = self.w('target')
+		assert self.find('target') == target
+
+	def test_parent(self):
+		"""
+		Single widget with a child.
+		"""
+		target = self.w('target')
+		self.w('foobar', [
+			target
+		])
+
+		assert self.find('target') == target
+
+	def test_hierarchy(self):
+		"""
+		Target widget with multiple parents and another widget with the same name, but
+		without parent.
+		"""
+		target = self.w('target')
+
+		self.w('bla', [
+			self.w('baz', [
+				target
+			])
+		])
+		self.w('foobar')
+		self.w('target')
+
+		assert self.find('bla/baz/target') == target
+
+	def test_hierarchy_incomplete_path(self):
+		"""
+		Target with multiple parents, select it while omitting one parent in the search
+		path.
+		"""
+		target = self.w('target')
+
+		self.w('bla', [
+			self.w('baz', [
+				target
+			])
+		])
+		self.w('foobar')
+		self.w('target')
+
+		assert self.find('bla/target') == target
```

Comment:
Having looked at the implementation now, this seems like a bigger problem to me, as I cannot provide a more specific path to `target2`. The only choice would be to rename some part. So at least an `exact=True` should be suppliable to `find()` for cases like this.